### PR TITLE
[GL-476] Docker Image Best Practices - Previously [GL-1379] Add AppSec trivy scans on PR

### PR DESCRIPTION
**Pull request by dinvlad**
This PR serves as a base branch for all of the child issues of epic **[GL-467](https://broadinstitute.atlassian.net/browse/GL-467)** This PR was originally created under the issue **[GL-1379](https://broadinstitute.atlassian.net/browse/GL-1379)** Add AppSec trivy scans by @dinvlad. 

The purpose of having this base branch is to create a single merge into develop after all of the docker changes have been made, this avoids having to run the scientific tests for multiple merges into develop for non-functional docker changes.
- @wjdingman1 

Child Issues being merged in:

- **[GL-1379](https://broadinstitute.atlassian.net/browse/GL-1379)** Add AppSec trivy scans
- **[GL-1567](https://broadinstitute.atlassian.net/browse/GL-1567)** Fix DSDE toolbox scanning
- **[GL-1590](https://broadinstitute.atlassian.net/browse/GL-1590)** Cleanup broad image scripts
- **[GL-1591](https://broadinstitute.atlassian.net/browse/GL-1591)** Update all broad images to python3
- **[GL-1575](https://broadinstitute.atlassian.net/browse/GL-1575)** Build and use samtools specific image
- **[GL-1598](https://broadinstitute.atlassian.net/browse/GL-1598)** Build and use GATK specific image
- **[GL-1604](https://broadinstitute.atlassian.net/browse/GL-1604)** Standardize formatting and style across dockerfiles
- **[GL-1599](https://broadinstitute.atlassian.net/browse/GL-1599)** Build and use SAMTOOLS + PICARD + BWA specific image
- **[GL-1592](https://broadinstitute.atlassian.net/browse/GL-1592)** Docker README
- **[GL-1593](https://broadinstitute.atlassian.net/browse/GL-1593)** Docker style guide
- **[GL-468](https://broadinstitute.atlassian.net/browse/GL-468)** Publish docker images to all specified locations

### CODE REVIEW

**Comment by dinvlad** [comment]
The `marketplace.gcr.io/google/debian9` image unfortunately cannot be accessed without Google credentials - would it be possible to use another base in that case?

**Comment by dinvlad** [comment]
Also, please let me know if there any other images/Dockerfiles you use in GP production.

**Comment by gbggrant** [comment]
Hi @dinvlad we specifically used the marketplace.gcr.io/google/debian9 managed docker image for security reasons.  Isn't it possible to set up a service account to be able to access those docker images?  We think that the security provided by google-generated docker images should be as good as our own.

File: dockers/broad/genomes_on_the_cloud/build_docker.sh
Comment (gbggrant) [inline dockers/broad/genomes_on_the_cloud/build_docker.sh:16]: Not on this line, but please change DOCKER_VERSION (line 10) to 2.4.8 (minor version change - should be functionally the same…)

**Review by gbggrant — COMMENTED** [review]
Looks good.  But we have some cleaning up to do.  So… Once this PR is merged, we have two weeks to fix them all?

File: .github/workflows/trivy.yml
Comment (jessicaway) [inline .github/workflows/trivy.yml:4]: Is it possible to configure this file so that we only see 1 github action that runs all these checks with an all or nothing pass/fail? The test section is already very long

Comment (dinvlad) [reply .github/workflows/trivy.yml:4]: I think the idea of a separate action was to not break your builds! But given that we've used this tool throughput many repos for several months now, this doesn't happen often. So may as well just add it as part of your overall workflow.

File: .github/workflows/trivy.yml
Comment (gbggrant) [reply .github/workflows/trivy.yml:26]: That sounds good.  To be clear, you'd put the prebuild-scan task into another PR, but that PR/ticket would be blocked until we deal with the dsde-toolbox and picard-cloud images?

Comment (wjdingman1) [reply .github/workflows/trivy.yml]: yeah, I moved this task out and created a separate branch off this one with it reintroduced which we can figure out next week what to do with it

**Comment by jessicaway** [comment]
add to whitelist

**Comment by wjdingman1** [comment]
@jessicaway Thought about it some more, I think making separate PRs for cemba and skylab would make more sense than to jam it all into this one. So the remaining work for this will just be to build&push the broad images and update them in the corresponding wdls....then will tackle picard, cemba skylab...thoughts?

**Comment by jessicaway** [comment]
@wjdingman1 that plan sounds good!

**Review by jessicaway — APPROVED** [review]

**Review by gbggrant — APPROVED** [review]

File: pipelines/broad/dna_seq/germline/single_sample/exome/ExomeGermlineSingleSample.changelog.md
Comment (jessicaway) [inline pipelines/broad/dna_seq/germline/single_sample/exome/ExomeGermlineSingleSample.changelog.md:1]: You can put these into one version. Multiple commits and changes are allowed for one version increase

File: pipelines/broad/dna_seq/germline/single_sample/wgs/WholeGenomeGermlineSingleSample.changelog.md
Comment (jessicaway) [inline pipelines/broad/dna_seq/germline/single_sample/wgs/WholeGenomeGermlineSingleSample.changelog.md:1]: Same comment here

File: pipelines/broad/dna_seq/germline/variant_calling/VariantCalling.changelog.md
Comment (jessicaway) [inline pipelines/broad/dna_seq/germline/variant_calling/VariantCalling.changelog.md:1]: and here

File: pipelines/broad/dna_seq/germline/single_sample/wgs/WholeGenomeGermlineSingleSample.changelog.md
Comment (jessicaway) [inline pipelines/broad/dna_seq/germline/single_sample/wgs/WholeGenomeGermlineSingleSample.changelog.md:10]: Did we actually change the BWA version? the version looks the same to me (but we will want to consult scientific owners if this does change the version of BWA)

**Review by jessicaway — COMMENTED** [review]
Looks good overall, just want to clarify the BWA change and then I will approve

File: dockers/broad/genomes_on_the_cloud/build_docker.sh
Comment (wjdingman1) [inline dockers/broad/genomes_on_the_cloud/build_docker.sh:18]: @jessicaway see here

File: pipelines/broad/dna_seq/germline/single_sample/wgs/WholeGenomeGermlineSingleSample.changelog.md
Comment (wjdingman1) [reply pipelines/broad/dna_seq/germline/single_sample/wgs/WholeGenomeGermlineSingleSample.changelog.md:10]: It was in the GoTC dockerfile...the release URL that was previously being referenced didn't actually exist. they had some weird version tag....i just rolled it down to 0.7.15 (https://github.com/lh3/bwa/releases/tag/v0.7.15)

probably will be good to double check...could still be the same underlying executable

**Comment by dinvlad** [comment]
Thanks a lot for working on these, I know it's a lot to be done!

File: pipelines/broad/dna_seq/germline/single_sample/wgs/WholeGenomeGermlineSingleSample.changelog.md
Comment (jessicaway) [reply pipelines/broad/dna_seq/germline/single_sample/wgs/WholeGenomeGermlineSingleSample.changelog.md:5]: Ok, yeah, that looks fine. I believe that is the same version of [BWA](https://github.com/lh3/bwa/releases)

**Comment by wjdingman1** [comment]
> Thanks a lot for working on these, I know it's a lot to be done!

Not a problem! Its a necessary step to add to our PRS!

File: dockers/broad/samtools/docker_build.sh
Comment (gbggrant) [inline dockers/broad/samtools/docker_build.sh:11]: I think this should be the samtools version primarily.  Probably important to have the docker image version here too i guess.

File: dockers/broad/samtools/docker_versions.tsv
Comment (gbggrant) [inline dockers/broad/samtools/docker_versions.tsv:2]: Same comment that I think this should be/include the samtools version (1.11) primarily - and then presumably the docker version and timestamp?  Maybe discuss.

**Review by gbggrant — APPROVED** [review]
Approve, but strongly suggest naming docker images with samtools version number (and use a similar naming convention for other tool-specific dockers).

File: dockers/broad/samtools/docker_build.sh
Comment (wjdingman1) [reply dockers/broad/samtools/docker_build.sh:11]: yeah i think thats totally reasonable, very easy fix!

File: dockers/broad/illumina_iaap_autocall/docker_build.sh
Comment (gbggrant) [inline dockers/broad/illumina_iaap_autocall/docker_build.sh:11]: Should include version number of iaap (1.1.0)

**Review by khajoue2 — APPROVED** [review]

**Comment by wjdingman1** [comment]
retest this please

**Pull request merged**
